File: Object/Shape/Quad.cpp
#include "Object/Shape/Quad.hpp"

#include <math.h>
#include <algorithm>

namespace Object {
    namespace Shape {
        Quad::Quad(const Math::Point &position, const Math::Vector &side1, const Math::Vector &side2)
            : mPosition(position), mSide1(side1), mSide2(side2)
        {
            mNormal = Math::Normal(mSide1 % mSide2).normalize();
            mTangent = Math::Bivector(mSide1, mSide2);
        }

        bool Quad::intersect(const Math::Ray &ray, Intersection &isect) const
        {
            float distance = ((ray.origin() - mPosition) * mNormal) / (ray.direction() * -mNormal);
            if (distance >= 0 && distance < isect.distance) {
                Math::Point point = ray.origin() + ray.direction() * distance;
                float u = (point - mPosition) * mSide1 / mSide1.magnitude2();
                float v = (point - mPosition) * mSide2 / mSide2.magnitude2();
                if (u >= 0 && u <= 1 && v >= 0 && v <= 1)
                {
                    isect.distance = distance;
                    isect.normal = mNormal;
                    isect.tangent = mTangent;
                    isect.surfacePoint = Math::Point2D(u, v);
                    return true;
                }
            }

            return false;
        }

        bool Quad::sample(Math::Sampler::Base &sampler, Math::Point &pnt, Math::Normal &nrm, float &pdf) const
        {
            Math::Point2D pntSurface = sampler.getValue2D();
            pnt = mPosition + mSide1 * pntSurface.u() + mSide2 * pntSurface.v();
            nrm = mNormal;
            
            float surfaceArea = (mSide1 % mSide2).magnitude();
            pdf = 1.0f / surfaceArea;

            return true;
        }

        float Quad::samplePdf(const Math::Point &) const
        {
            float surfaceArea = (mSide1 % mSide2).magnitude();
            return 1.0f / surfaceArea;
        }

        BoundingVolume Quad::boundingVolume(const Math::Transformation &trans) const
        {
            Math::Point points[] = { mPosition, mPosition + mSide1, mPosition + mSide2, mPosition + mSide1 + mSide2 };

            BoundingVolume volume;
            for (const Math::Point &point : points) {
                volume.expand(trans * point);
            }

            return volume;
        }

        void Quad::writeProxy(ShapeProxy &proxy) const
        {
            proxy.type = ShapeProxy::Type::Quad;
            mPosition.writeProxy(proxy.quad.position);
            mSide1.writeProxy(proxy.quad.side1);
            mSide2.writeProxy(proxy.quad.side2);
            mNormal.writeProxy(proxy.quad.normal);
        }
    }
}

File: src/object/shape/quad.rs
use crate::geo;
use crate::object;

use geo::Bivec3;
use geo::Normal3;
use geo::Point2;
use geo::Point3;
use geo::Ray;
use geo::Transformation;
use geo::Vec3;

use object::BoundingVolume;
use object::Sampler;
use object::Shape;
use object::ShapeIntersection;

pub struct Quad {
    position : Point3,
    side1 : Vec3,
    side2 : Vec3,
    normal : Normal3
}

impl Quad {
    pub fn new(position : Point3, side1 : Vec3, side2 : Vec3) -> Quad {
        let normal = Normal3::from((side1 % side2).normalize());
        Quad{position, side1, side2, normal}
    }
}

impl Shape for Quad {
    fn intersect(&self, ray : Ray, max_distance : f32, _closest : bool) -> Option<ShapeIntersection> {
        let distance = (ray.origin - self.position) * self.normal / (ray.direction * -self.normal);
        
        if distance >= 0.0 && distance < max_distance {
            let point = ray.origin + ray.direction * distance;
            let u = (point - self.position) * self.side1 / self.side1.mag2();
            let v = (point - self.position) * self.side2 / self.side2.mag2();
            if u >= 0.0 && u <= 1.0 && v >= 0.0 && v <= 1.0 {
                return Some(ShapeIntersection::new(distance, self.normal, Bivec3::new(self.side1, self.side2), Point2::new(u, v)));
            }
        }

        return None;
    }

    fn bounding_volume(&self, xform : Transformation) -> BoundingVolume {
        let mut volume = BoundingVolume::new();
        for point in [self.position, self.position + self.side1, self.position + self.side2, self.position + self.side1 + self.side2] {
            volume.include_point(point.transform(xform));
        }

        return volume;
    }

    fn sample(&self, sampler : &mut dyn Sampler) -> Option<(Point3, Normal3, f32)> {
        let pnt_surface = sampler.get_value2();
        let pnt = self.position + self.side1 * pnt_surface.u + self.side2 * pnt_surface.v;
        let nrm = self.normal;

        let surface_area = (self.side1 % self.side2).mag();
        let pdf = 1.0 / surface_area;

        return Some((pnt, nrm, pdf));        
    }

    fn sample_pdf(&self, _pnt : Point3) -> f32 {
        let surface_area = (self.side1 % self.side2).mag();
        return 1.0 / surface_area;
    }
}